### PR TITLE
Extract/jetpack plugin installer

### PR DESCRIPTION
**Pull request by leogermani**
This PR extracts the Jetpack_Plugins class into its own package.

In the future we might want to extract the REST API endpoints that use it to the package as well.

The approach was to keep the same class names and add the new package as a dependency to the Jetpack plugin. Doing that, the autoloader will take care to include the class whenever it's needed.

#### Changes proposed in this Pull Request:

* extracts the Jetpack_Plugins class into its own package.

#### Jetpack product discussion

p1643119059163700-slack-CBG1CP4EN

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* Set up a new Jurassic Ninja site with this branch
* Activate and connect Jetpack
* Add a new post
* Add a Contact Form block
* On the block options panel, go to "Newsletter Integration"
* Click to install the Creative Email plugin
* Make sure the plugin is properly installed and activated

The above tests the sideloading of plugins. Now let's test the REST endpoint that allows user to upload a zip file to add plugins to their site

* Go to wordpress.com
* Switch site to your testing Jetpack site
* Go to Plugins
* Click "Upload" on the top of the screen
* Upload a zip file with any valid plugin
* Go back to your site's wp-admin and verify that the plugin was installed and activated.

Now let's test the backward compatibility of the changes. Let's open a terminal and do things the old way.

Open a terminal with `wp shell`

* `jetpack_require_lib( 'plugins' );` should return null
* Install a plugin using the old class, for example woocommerce:
* `Jetpack_Plugins::install_plugin( 'woocommerce' );`
* Verify that the plugin is installed

### CODE REVIEW

File: projects/packages/plugins-installer/src/class-jetpack-automatic-install-skin.php
Comment (anomiex) [inline projects/packages/plugins-installer/src/class-jetpack-automatic-install-skin.php:62]: Seems like there could be a better description for this function.
```suggestion
	 * Set the main error code.
	 *
	 * Don't set the process_failed as code since it is not that helpful unless we don't have one already set
```

File: projects/packages/plugins-installer/src/class-jetpack-automatic-install-skin.php
Comment (anomiex) [inline projects/packages/plugins-installer/src/class-jetpack-automatic-install-skin.php:72]: Seems like there could be a better description for this function.
```suggestion
	 * Set the main error message.
	 *
	 * Don't set the process_failed as message since it is not that helpful unless we don't have one already set
```

**Review by anomiex — COMMENTED** [review]
Gave it a once-over from a monorepo perspective.

It'd be nice if we could rename the class to something like `Automattic\Jetpack\PluginsInstaller` if there aren't any external uses (as suggested by p1643119495165100/1643119059.163700-slack-CBG1CP4EN).

File: projects/packages/plugins-installer/src/class-jetpack-automatic-install-skin.php
Comment (leogermani) [reply projects/packages/plugins-installer/src/class-jetpack-automatic-install-skin.php:83]: I think the code is right. It checks for the error code but sets the error message.

**Comment by leogermani** [comment]
> Gave it a once-over from a monorepo perspective.
> 
> It'd be nice if we could rename the class to something like `Automattic\Jetpack\PluginsInstaller` if there aren't any external uses (as suggested by p1643119495165100/1643119059.163700-slack-CBG1CP4EN).

Yes, it definitely would. But I don't see a way of doing this in a fully backward compatible manner, unless we also kept the deprecated classes in the Jetpack plugin, leaving more garbage behind... what do you think?

I just re-added the removed files, leaving them empty with just a comment. We could leave the full classes and add deprecation warnings to all methods - and then replace all usages of it to the new classes...

File: projects/packages/plugins-installer/src/class-jetpack-automatic-install-skin.php
Comment (anomiex) [reply projects/packages/plugins-installer/src/class-jetpack-automatic-install-skin.php:83]: I haven't tried it, but it looks to me that with the current code you'd get the following behavior:
```php
$instance->set_main_error_code( "foo" );
$instance->set_main_error_message( "Message for foo", "foo" );
echo $instance->get_main_error_message() . "\n"; // Prints "Message for foo".
$instance->set_main_error_message( "Unused message", "process_failed" );
echo $instance->get_main_error_message() . "\n"; // Now prints "foo", not "Message for foo".
```
Whether we should check the code or message in the first part is debatable (personally I'd probably merge `set_main_error_code` and `set_main_error_message` into one method to avoid the ambiguity), but the second should certainly not be the code.

**Comment by anomiex** [comment]
The stub files could do like `class_alias( Automattic\Jetpack\PluginsInstaller::class, Jetpack_Plugins::class );` to maintain compat.

**Comment by leogermani** [comment]
Great idea @anomiex 

I've done that and updated the test instructions to test the backward compatibility of the changes.

File: projects/packages/plugins-installer/src/class-automatic-install-skin.php
Comment (retrofox) [inline projects/packages/plugins-installer/src/class-automatic-install-skin.php:38]: minor: typo

**Review by retrofox — COMMENTED** [review]
Testing the functionality of the implementation according to instructions.

I've added some error_log lines to follow the installing plugin process:

**Getting the plugins list**


**Installing a plugin**

https://user-images.githubusercontent.com/77539/151193505-56014a97-454f-457e-85c8-8590cfd04c6c.mov





**Activating plugin**




It works as expected, too.

**Review by retrofox — COMMENTED** [review]
wpcom side


https://user-images.githubusercontent.com/77539/151196206-df295aee-3ef3-4d19-b956-ab487a3259d6.mov


I didn't see any activity in the testing site backend, though :-/ Maybe doing something wrong?

**Review by retrofox — APPROVED** [review]
Accepting ✅ but I suggest another couple of 👀 here before merging it.

File: projects/packages/plugins-installer/src/class-jetpack-automatic-install-skin.php
Comment (leogermani) [reply projects/packages/plugins-installer/src/class-jetpack-automatic-install-skin.php]: Yeah I get your point now. I'm not sure if we should change the second check against the existence of the code or the message...

I went ahead and merged the PRto unblock the work and will prepare a new one to address this.

Worth saying that this has been there for many many years, I just moved the code around.

**Pull request merged**
